### PR TITLE
Throw exception in case of WAL failure instead of only printing a message

### PR DESCRIPTION
**Pull request by Mytherin**
This PR reworks the way that failures during WAL playback are handled. Exceptions during WAL playback can occur when dealing with unclean exits - in particular, we can encounter serialization exceptions when the WAL is truncated due to a previous failure. 

However, non-serialization exceptions are not expected here, and generally means something has gone wrong (either a bug in the code, the WAL file does not belong to the database file, etc). In this case, we would previously print this message but continue operations.

After this PR, we instead throw on all non-serialization exceptions. In addition, we add a new setting: `abort_on_wal_failure`. If this setting is true, we **always** abort on an exception during WAL playback, even if the exception is a serialization exception. This setting defaults to false.

### CODE REVIEW

**Comment by Tishj** [comment]
Is the idea that we add this to the SQLLogicTestRunner, enabling this setting by default there?

File: src/storage/wal_replay.cpp [hunk 1]
```diff
@@ -235,17 +231,19 @@ bool WriteAheadLog::Replay(AttachedDatabase &database, unique_ptr<FileHandle> ha
...
 		}
 	} catch (std::exception &ex) { // LCOV_EXCL_START
+		// exception thrown in WAL replay: rollback
+		con.Query("ROLLBACK");
 		ErrorData error(ex);
-		if (error.Type() != ExceptionType::SERIALIZATION) {
-			// FIXME: this should report a proper warning in the connection
-			Printer::PrintF("Exception in WAL playback: %s\n", error.RawMessage());
-			// exception thrown in WAL replay: rollback
+		// serialization failure means a truncated WAL
```
Comment (carlopi) [inline src/storage/wal_replay.cpp:237]: Unsure what guarantees this, now or in the future, I think you can probably get a SerializationException from other cases (unsure how, but we throw SerializationException in many places), even though I don't see exactly a path to this happening.

**Comment by Mytherin** [comment]
@taniabogatsch the failure in this PR seems to be caused by https://github.com/duckdb/duckdb/pull/12049
 - but as we swallowed WAL errors the issue was ignored. See the CI logs where the message is printed - https://github.com/duckdb/duckdb/actions/runs/9091829925/job/24998220854#step:7:901.

The problem seems to be caused by the WAL not being correctly deleted after a checkpoint, as `storage_manager.GetWALSize()` returns 0, despite a WAL file being there physically on disk.

Perhaps you could have a look?

**Comment by taniabogatsch** [comment]
Have a look at my requested PR: https://github.com/Mytherin/duckdb/pull/11. I think this happens in the rare case that we use `GetWALSize` while the WAL is not yet open. The test also combines `--force-storage --force-reload` with `PRAGMA disable_checkpoint_on_shutdown;`, causing a rare combination of flags. So we DO checkpoint, just not on shutdown.

**Pull request merged**